numpy/core/src/multiarray/arrayfunction_override.c

```diff
@@ -329,6 +336,93 @@ array_implement_array_function(
 }
 
 
+/*
+ * Implements the __array_function__ protocol for a Python function, as described in
+ * in NEP-18. See numpy.core.overrides for a full docstring.
+ */
+NPY_NO_EXPORT PyObject *
+array_implement_array_function(
+    PyObject *NPY_UNUSED(dummy), PyObject *positional_args)
+{
+    PyObject *implementation, *public_api, *relevant_args, *args, *kwargs;
+
+    if (!intern_strings())
+        return NULL;
+
+    if (!PyArg_UnpackTuple(
+            positional_args, "implement_array_function", 5, 5,
+            &implementation, &public_api, &relevant_args, &args, &kwargs))
+        return NULL;
+
+    /* Remove `like=` kwarg, which is NumPy-exclusive and thus not present
+     * in downstream libraries.
+     */
+    if (PyDict_CheckExact(kwargs) &&
+            PyDict_Contains(kwargs, npy_arrayfunction_str_like))
+        PyDict_DelItem(kwargs, npy_arrayfunction_str_like);
+
+    return array_implement_array_function_internal(
+        implementation, public_api, relevant_args, args, kwargs);
+}
+
+
+/*
+ * Implements the __array_function__ protocol for a C function. Added as an extension
+ * to NEP-18 in an effort to bring NEP-35 to life with minimal dispatch overhead.
+ */
+NPY_NO_EXPORT PyObject *
+array_implement_c_array_function(
+    const char *function_name, PyObject *args, PyObject *kwargs)
+{
+    PyObject *relevant_args;
+    PyObject *numpy_module = NULL, *public_api = NULL;
+    PyObject *result = NULL;
+
+    if (kwargs == NULL)
+        return NULL;
+
+    if (!intern_strings())
+        goto cleanup;
+
+    /* Remove `like=` kwarg, which is NumPy-exclusive and thus not present
+     * in downstream libraries. If that key isn't present, return NULL and
+     * let originating call to continue.
+     */
+    if (PyDict_CheckExact(kwargs) &&
+            PyDict_Contains(kwargs, npy_arrayfunction_str_like)) {
+        relevant_args = PyTuple_Pack(1,
+                PyDict_GetItem(kwargs, npy_arrayfunction_str_like));
+        if (relevant_args == NULL)
+            return NULL;
+        PyDict_DelItem(kwargs, npy_arrayfunction_str_like);
+    }
+    else {
+        return NULL;
+    }
+
+    numpy_module = PyImport_Import(npy_arrayfunction_str_numpy);
```

Comment:
Yes, I thought of doing that and decided not to do it for now because it would add quite some complexity to the code (e.g., a hash table that supports both adding and searching on strings) that is _probably_ not worth it given that the downstream library will most likely dominate time anyway. It's still something we could improve in the future if it implies a non-negligible performance impact.

Reply:
I think the solution would be to pass in a pointer to a `static PyObject *` in the calling function as additional argument, but I agree its not worth the trouble, especially since this is already in the non-NumPy side.

Reply:
That would be a simple alternative I didn't think of initially. I think this is definitely something we could consider, the downside is it would pollute the calling code a bit and you would definitely pay the cost of importing each of the functions at least once, which is probably not something we would like to do. I think something on the `arrayfunction_override.c` side would be preferable to avoid such nuisances, but feel free to correct me if I'm wrong.